### PR TITLE
Allow ISignatureTypeProvider to be a struct

### PR DESCRIPTION
**Pull request by terrajobst**
The `SignatureDecoder` takes the provider as a naked interface. This makes it inefficient to bake it by a struct because it will cause it being boxed each time it's passed.

However, it's very desirable to make the provider a struct because it allows the caller to provide call site specific information.

For example, in order for implementers to reference the generic type parameters of the method or type, the `ISignatureTypeProvider` needs to have access to the containing type / containing method.

To avoid the boxing operation, the methods simply needs an additional contraint like so:

```csharp
public static TType DecodeType<TType, TProvider>(Handle handle, TProvider provider)
    where TProvider : ISignatureTypeProvider<TType>
```

Unfortunately, this causes some noise as C# is no longer able to infer the type arguments, so each call site must now qualify the type arguments.

### CODE REVIEW

**Comment by terrajobst** [comment]
@nguerrera @davkean

**Comment by terrajobst** [comment]
Hmm, the test failures seem to be unrelated to my changes. It looks the tests are unable to find `System.Collections.Immutable.dll`.

**Comment by stephentoub** [comment]
@nguerrera, this failure looks similar to what you fixed with #1026...

File: src/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj [hunk 1]
```diff
@@ -13,10 +13,14 @@
     <TargetFrameworkProfile>Profile7</TargetFrameworkProfile>
     <FileAlignment>512</FileAlignment>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <SccProjectName>SAK</SccProjectName>
-    <SccLocalPath>SAK</SccLocalPath>
-    <SccAuxPath>SAK</SccAuxPath>
-    <SccProvider>SAK</SccProvider>
+    <SccProjectName>
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System.Reflection.Metadata.csproj]: Noise. These should be deleted altogether, but in a separate, low-pri change.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/SignatureDecoder.cs [hunk 2]
```diff
@@ -17,7 +16,8 @@ public static class SignatureDecoder
         /// <param name="handle">A type definition, reference, or specification handle.</param>
         /// <param name="provider">The type provider.</param>
         /// <exception cref="System.BadImageFormatException">The handle does not represent a valid type reference, definition, or specification.</exception>
-        public static TType DecodeType<TType>(Handle handle, ISignatureTypeProvider<TType> provider)
+        public static TType DecodeType<TType, TProvider>(Handle handle, TProvider provider)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Decoding/SignatureDecoder.cs]: We pass this provider around liberally. Should it be passed by ref everywhere if we now think it should be a struct?

**Comment by nguerrera** [comment]
@dotnet-bot test this please

**Comment by nguerrera** [comment]
@terrajobst @stephentoub I've merged master to dev/metadata, which should address the CI failure.

**Comment by nguerrera** [comment]
cc @tmat

@terrajobst This is interesting, but the loss of type inference is unfortunate and it does make the API harder to understand IMHO. With that said, perf is a very explicit goal of the metadata reader and as you demonstrated in your sample offline, the typical caller has only a few call sites straight to this API. 

I would note that decoding does allocate immutable arrays, e.g. for parameter types. We've said so far that this API is OK to be a bit more heavyweight than the lower level parts. Even Roslyn's rough equivalent MetadataDecoder's are not structs and are allocated to retain the same context you describe in the PR description.

**Comment by tmat** [comment]
What about making SignatureDecoder a generic struct with TType and TProvider parameters instead of a static class?

var decoder = new SignatureDecoder<TypeSymbol, SignatureProvider>(myProvider);

decoder.DecoderType(handle)

**Comment by stephentoub** [comment]
@tmat, @nguerrera, @terrajobst, what's the status of this PR?

**Comment by tmat** [comment]
@stephentoub These changes are in vNext version, which is currently not getting much attention due to other priorities. I think we should make SignatureDecoder a generic struct.

**Comment by stephentoub** [comment]
Thanks, @tmat.  I realize it's against dev/metadata, I just worry a bit when I see a PR that's been sitting  open against a main branch without visible progress for almost two months.  If there's a design question outstanding, makes me question whether it should be an issue instead of a PR until the design issues are worked out and a PR can be submitted that matches it.  If you'd prefer to just leave this open, though, up to you.

**Comment by nguerrera** [comment]
I'm going to do as Tomas suggested and submit it as a separate PR.

**Pull request closed without merging**